test/Microsoft.AspNetCore.Routing.Tests/RouteTest.cs

```diff
@@ -647,6 +650,37 @@ public void GetVirtualPath_EncodesValues()
             Assert.Empty(pathData.DataTokens);
         }
 
+        [Fact]
+        public void GetVirtualPath_AlwaysUsesDefaultUrlEncoder()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            services.AddSingleton<ILoggerFactory>(NullLoggerFactory.Instance);
+            services.AddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
+            services.AddRouting();
+            // This test encoder should not be used by Routing and should always use the default one.
+            services.AddSingleton<UrlEncoder>(new UrlTestEncoder());
+            var httpContext = new DefaultHttpContext
+            {
+                RequestServices = services.BuildServiceProvider(),
+            };
+
+            var context = new VirtualPathContext(
+                httpContext,
+                values: new RouteValueDictionary(new { name = "name with %special #characters" }),
```

Comment:
Could we add a test for the actual issue that was reported? I don't think the reported problem was with `%` and `#`